### PR TITLE
Tidy up MKL configuration.

### PR DESCRIPTION
**Pull request by gunan**
Also only use "--config=mkl" to build with MKL support.

### CODE REVIEW

**Comment by gunan** [comment]
@tfboyd Could you get reviewers from Intel to look at this PR? Also, if we have any documentation for mkl, it will need to be updated..

@allenlavoie @jart Looks like my bazel configs were correct, but build_pip_package script had some hardcoded references to mkl libraries.

**Comment by gunan** [comment]
@tfboyd, this PR blocks my further work on having MKL on windows and macos.
However, this PR should be sufficient to build TF with both MKL support and cuda.

**Comment by gunan** [comment]
windows cmake tests have an issue, but they are known issues due to a bug in `tf.acosh`

File: tensorflow/core/kernels/mkl_tfconv_op.cc [hunk 1]
```diff
@@ -29,8 +29,8 @@ limitations under the License.
 #include "tensorflow/core/util/tensor_format.h"
 
 #include "tensorflow/core/util/mkl_util.h"
-#include "third_party/mkl/include/mkl_dnn.h"
-#include "third_party/mkl/include/mkl_dnn_types.h"
+#include "mkl_dnn.h"
```
Comment (drpngx) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: Why this change?

Comment (gunan) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: previously all the headers were manually copied to third_paarty/mkl/include.
However, now we use "includes" field of the cc_library rule "intel_binary_blob", which adds the include folder for mkl to the gcc commands as a "-iquote" flag. Then all these libraries become available to include without the need for a relative path.

Comment (Wang-Yue) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: Don't do that. This will cause various issues. Always use full include path in tensorflow code please.

Comment (gunan) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: Could you elaborate on 'various issues'?
It worked as is before, because of all the non-standard things that were done in configure. This is what we do for any other external dependency headers, libjpeg, libpng, etc. The path here was just a hand fabricated path due to the way things were copied in `confiure`

Comment (Wang-Yue) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: In Google projects, always use full include path for C++/ObjC programs if you can. That's the convention. 
If you have no other options, there's a couple of things you can do
- guard the full path inclusion for google use (PLATFORM_GOOGLE), and partial path for others. 
- in addition, you should use <lib.h> rather than "lib.h" whenever possible.

Comment (gunan) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: That convention is for internal includes. Anything that is a part of tensorflow. If something is an external dependency, we point our build rules to their include folder:
https://github.com/tensorflow/tensorflow/blob/master/third_party/png.BUILD#L31

Then include them directly in our files. This is what I am doing here. Getting rid of brittle downloading through configure script mechanism end up moving us to here.
Unless we check in the headers into our third_party/mkl folder, we cannot make them accessible through the `third_party/mkl/include`path.

Moving everything into a single header is something I had in mind, but that will come in a later PR.

File: tensorflow/tools/pip_package/build_pip_package.sh [hunk 2]
```diff
@@ -17,6 +17,10 @@
 
 set -e
 
+function real_path() {
```
Comment (drpngx) [inline tensorflow/tools/pip_package/build_pip_package.sh]: I think `readlink -f` should work on modern systems.

File: tensorflow/core/kernels/mkl_tfconv_op.cc (hunk 1 above)
Comment (drpngx) [inline tensorflow/core/kernels/mkl_tfconv_op.cc]: Gotcha.

File: third_party/mkl/mkl.BUILD [hunk 3]
```diff
@@ -0,0 +1,22 @@
+licenses(["restricted"])  # MPL2, portions GPL v3, LGPL v3, BSD-like     TODO
```
Comment (drpngx) [inline third_party/mkl/mkl.BUILD]: The web site says Apache... Could you check or get someone to check?

**Comment by agramesh1** [comment]
One reason we had added MKL download in the config file was to enable the user to be able to point to point to a local copy instead of downloading it. Is it possible to accomplish the same if the download is  through bazel? 

Other than this the changes look fine.

**Comment by gunan** [comment]
@jart, how can we enable users to use their own copies of MKL with this setup?

@agramesh1 Can you provide input on the license question @drpngx added on the mkl.BUILD file?

File: third_party/mkl/mkl.BUILD (hunk 3 above)
Comment (agramesh1) [inline third_party/mkl/mkl.BUILD]: MKL ML is distributed under 3-clause BSD.

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
```
FAIL: mismatch in packaged licenses and external dependencies
Missing the licenses for the following external dependencies:
//third_party/mkl
@mkl//
```

**Comment by drpngx** [comment]
(Removing "awaiting testing then merge", since @jart might have some input)

File: tensorflow/tools/pip_package/build_pip_package.sh (hunk 2 above)
Comment (gunan) [inline tensorflow/tools/pip_package/build_pip_package.sh]: When I try that on macOS, I see this:
```
gunan$ readlink -f .
readlink: illegal option -- f
usage: readlink [-n] [file ...]
```

readlink -f seems to be OK with GNU coreutils, but it does not seem to work with BSD coreutils.

**Comment by gunan** [comment]
Resolved license issue. 
I will tidy things up further when adding macOS support by removing third_party/mkl directory.

@agramesh1 looks like as @drpngx pointed out the license in MKL girhub repository is apache.
You might want to update the license there.

**Comment by gunan** [comment]
@jart do you have any comments on BUILD files.

**Comment by gunan** [comment]
Ping! Any objections to this PR?

**Comment by agramesh1** [comment]
@gunan  I will get MKL team to update license on github repository.  Regarding the PR,  we want be sure there is a way for users to use their own copy of MKL. Awaiting response from @jart

**Comment by gunan** [comment]
Current PR removes straightforward support for working with your local copy of MKL.
If that is a vital feature you need, how about I add that with a subsequent PR?

File: tensorflow/tools/pip_package/build_pip_package.sh (hunk 2 above)
Comment (jart) [inline tensorflow/tools/pip_package/build_pip_package.sh]: Why do we need a realpath function?

File: tensorflow/workspace.bzl [hunk 4]
```diff
@@ -172,6 +172,17 @@ def tf_workspace(path_prefix="", tf_repo_name=""):
   )
 
   native.new_http_archive(
+      name = "mkl",
+      urls = [
+          "https://github.com/01org/mkl-dnn/releases/download/v0.7/mklml_lnx_2018.0.20170425.tgz",
```
Comment (jart) [inline tensorflow/workspace.bzl]: Mirrored.

File: tensorflow/workspace.bzl [hunk 5]
```diff
@@ -172,6 +172,17 @@ def tf_workspace(path_prefix="", tf_repo_name=""):
...
+          "http://mirror.bazel.build/github.com/01org/mkl-dnn/releases/download/v0.7/mklml_lnx_2018.0.20170425.tgz",
```
Comment (jart) [inline tensorflow/workspace.bzl]: The downloader is designed to *always* try the first URL in the list, and then begins to fail over in parallel to the second URL after like 2000ms. So the fastest URL should always be the first URL.

Comment (gunan) [inline tensorflow/workspace.bzl]: Done

File: tensorflow/workspace.bzl (hunk 4 above)
Comment (gunan) [inline tensorflow/workspace.bzl]: thank you very much!

**Comment by gunan** [comment]
real_path() addition is sort of irrelevant to this change, but when experimenting I discovered a bug in build pip package. If I ran build_pip_package script with a relative path, the whl file just got lost. I wanted to fix that in this PR.
Would you like me to move it to a new PR?

**Comment by gunan** [comment]
I have prepared the local mkl configuration here:  https://github.com/gunan/tensorflow/tree/local_mkl
Once this PR is merged, I will create a new pull request.
@agramesh1 the model I have will be this:
-configure script won't ask anything to the users about MKL.
-using `--config=mkl` will be sufficient to build a pip package with MKL support.
-If you use the environment variable `TF_MKL_ROOT`, `--config=mkl` will use that mkl instead of downloading it from the internet.
Does the above sound good to you?

**Comment by agramesh1** [comment]
@gunan Yes, that will work. Thanks.

**Comment by jhseu** [comment]
As discussed with @gunan, will submit internally to make this easier to merge. Closing.
